extensions/pkg/controller/heartbeat/controller.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package heartbeat
+
+import (
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// ControllerName is the name of the controller.
+const ControllerName = "heartbeat"
+
+// AddArgs are arguments for adding a heartbeat controller to a manager.
+type AddArgs struct {
+	// ControllerOptions are the controller.Options.
+	ControllerOptions controller.Options
+	// ExtensionName is the name of the extension controller.
+	ExtensionName string
+	// Namespace is the namespace which will be used for the heartbeat lease resource.
+	Namespace string
+	// RenewIntervalSeconds defines how often the heartbeat lease is renewed.
+	RenewIntervalSeconds int32
+	// Clock is the clock to use when renewing the heartbeat lease resource.
+	Clock clock.Clock
+}
+
+// Add creates a new heartbeat controller and adds it to the given manager.
+func Add(mgr manager.Manager, args AddArgs) error {
+	args.ControllerOptions.Reconciler = NewReconciler(args.ExtensionName, args.Namespace, args.RenewIntervalSeconds, args.Clock)
+	args.ControllerOptions.RecoverPanic = true
+
+	ctrl, err := controller.New(ControllerName, mgr, args.ControllerOptions)
+	if err != nil {
+		return err
+	}
+
+	eventChannel := make(chan event.GenericEvent, 1)
+	eventChannel <- event.GenericEvent{}
+
+	return ctrl.Watch(
+		&source.Channel{Source: eventChannel},
+		&handler.Funcs{GenericFunc: func(_ event.GenericEvent, q workqueue.RateLimitingInterface) { q.Add(reconcile.Request{}) }},
+	)
```

Comment:
To make this even simpler, you could use this: https://github.com/gardener/gardener/blob/48cb8af2b040e57136d5e317367d7e2656caa3d0/extensions/pkg/webhook/certificates/reconciler.go#L260-L264

Probably it makes sense to move this `triggerOnce` function to the `controllerutils` package like this:

```golang
// TriggerOnce is a source.Source that simply triggers the reconciler once with an empty reconcile.Request.
var TriggerOnce = source.Func(func(_ context.Context, _ handler.EventHandler, q workqueue.RateLimitingInterface, _ ...predicate.Predicate) error {
	q.Add(reconcile.Request{})
	return nil
})
```

```suggestion
	return ctrl.Watch(controllerutils.TriggerOnce, nil)
```